src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs

```diff
@@ -0,0 +1,667 @@
+            else
+            {
+                var newRoot = (element, priority);
+                _nodes[RootIndex] = newRoot;
+
+                MoveDown(newRoot, RootIndex);
+                _version++;
+
+                return root.element;
+            }
+        }
+
+        /// <summary>
+        /// Enqueues a collection of element/priority pairs.
+        /// </summary>
+        public void EnqueueRange(IEnumerable<(TElement Element, TPriority Priority)> items)
+        {
+            if (items is null)
+            {
+                throw new ArgumentNullException(nameof(items));
+            }
+
+            foreach (var (element, priority) in items)
+            {
+                Enqueue(element, priority);
+            }
+        }
+
+        /// <summary>
+        /// Enqueues a collection of elements, each associated with the specified priority.
+        /// </summary>
+        public void EnqueueRange(IEnumerable<TElement> elements, TPriority priority)
+        {
+            if (elements is null)
+            {
+                throw new ArgumentNullException(nameof(elements));
+            }
+
+            foreach (var element in elements)
+            {
+                Enqueue(element, priority);
+            }
+        }
+
+        /// <summary>
+        /// Removes all items from the priority queue.
+        /// </summary>
+        public void Clear()
+        {
+            if (RuntimeHelpers.IsReferenceOrContainsReferences<(TElement, TPriority)>())
+            {
+                // Clear the elements so that the gc can reclaim the references
+                Array.Clear(_nodes, 0, _size);
+            }
+            _size = 0;
+            _version++;
+        }
+
+        /// <summary>
+        /// Ensures that the priority queue has the specified capacity
+        /// and resizes its underlying array if necessary.
+        /// </summary>
+        public int EnsureCapacity(int capacity)
+        {
+            if (capacity < 0)
+            {
+                throw new ArgumentOutOfRangeException(
+                    nameof(capacity), capacity, SR.ArgumentOutOfRange_NeedNonNegNum);
+            }
+
+            int currentCapacity = _nodes.Length;
+            if (currentCapacity >= capacity)
+            {
+                return currentCapacity;
+            }
+
+            SetCapacity(capacity);
+            return capacity;
+        }
+
+        /// <summary>
+        /// Sets the capacity to the actual number of items in the priority queue,
+        /// if that is less than 90 percent of current capacity.
+        /// </summary>
+        public void TrimExcess()
+        {
+            int threshold = (int)(((double)_nodes.Length) * 0.9);
+            if (_size < threshold)
+            {
+                Array.Resize(ref _nodes, _size);
+                _version++;
+            }
+        }
+
+        private void EnsureEnoughCapacityBeforeAddingNode()
+        {
+            if (_size == _nodes.Length)
```

Comment:
I think its clearer to start with
```c#
if (_size < _nodes.Length)
    return;

... cont..
```
It is nice to reduce indentation  as it can make things easier to read.

It may mean the first branch may be predicted better, if the CPU initially assumes forward branches are not taken, but that isn't my area, and it surely doesn't matter here..

Reply:
Actually, I would take this check and move it out into Enqueue. That way in the common case, no method call is required. And this method can be renamed Grow()